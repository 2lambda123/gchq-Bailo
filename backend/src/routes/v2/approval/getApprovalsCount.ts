import bodyParser from 'body-parser'
import { Request, Response } from 'express'

interface GetApprovalsCountResponse {
  data: {
    count: number
  }
}

export const getApprovalsCount = [
  bodyParser.json(),
  async (req: Request, res: Response<GetApprovalsCountResponse>) => {
    return res.json({
      data: {
        count: 2,
      },
    })
  },
]