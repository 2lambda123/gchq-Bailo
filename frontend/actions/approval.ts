import qs from 'querystring'
import useSWR from 'swr'

import { ApprovalRequestInterface } from '../types/types'
import { ErrorInfo, fetcher } from '../utils/fetcher'

export function useGetApprovalRequestsForUser(isActive = true) {
  const { data, error, mutate } = useSWR<
    {
      data: { approvals: ApprovalRequestInterface[] }
    },
    ErrorInfo
  >(
    `/api/v2/approvals?${qs.stringify({
      isActive,
    })}`,
    fetcher
  )

  return {
    mutateApprovals: mutate,
    approvals: data ? data.data.approvals : [],
    isApprovalsLoading: !error && !data,
    isApprovalsError: error,
  }
}

export function useGetNumApprovals() {
  const { data, error, mutate } = useSWR<
    {
      data: { count: number }
    },
    ErrorInfo
  >('/api/v2/approvals/count', fetcher)

  return {
    mutateNumApprovals: mutate,
    numApprovals: data?.data.count,
    isNumApprovalsLoading: !error && !data,
    isNumApprovalsError: error,
  }
}
