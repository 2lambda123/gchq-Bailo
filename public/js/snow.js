const TOTAL_SNOWFLAKES = 200

let snowDiv = document.getElementById('snowContainer')

function embRand(a, b) {
  return Math.floor(Math.random() * (b - a + 1)) + a
}

if (!snowDiv) {
  const snowflakes = new Array(TOTAL_SNOWFLAKES).fill(`<i class="snow"></i>`).join('\n')

  const snow = `
    .snow {
      position: absolute;
      width: 10px;
      height: 10px;
      background: white;
      border-radius: 50%;
      margin-top:-10px;
    }

    #snowContainer {
      position: fixed;
      left: 0;
      top: 0;
      bottom: 0;
      width: 100vw;
      height: 100vh;
      overflow: hidden;
      z-index: 9999999;
      pointer-events: none;
    }
  `

  const transitions = new Array(TOTAL_SNOWFLAKES)
    .fill(undefined)
    .map((_, i) => {
      const rndX = embRand(0, 1000000) * 0.0001
      const rndO = embRand(-100000, 100000) * 0.0001
      const rndT = (embRand(3, 8) * 10).toFixed(2)
      const rndS = (embRand(0, 10000) * 0.0001).toFixed(2)

      return `
      .snow:nth-child(${i}) {
        opacity: ${(embRand(1, 10000) * 0.0001).toFixed(2)};
        transform: translate(${rndX.toFixed(2)}vw,-10px) scale(${rndS});
        animation: fall-${i} ${embRand(10, 30)}s -${embRand(0, 30)}s linear infinite;
      }

      @keyframes fall-${i}{
        ${rndT}%{
          transform: translate(${(rndX + rndO).toFixed(2)}vw,${rndT}vh) scale(${rndS});
        } to {
          transform: translate(${(rndX + rndO / 2).toFixed(2)}vw, 105vh) scale(${rndS});
        }
      }
    `
    })
    .join('\n')

  snowDiv = document.createElement('div')
  snowDiv.id = 'snowContainer'

  snowDiv.innerHTML = `
    <style>
      ${snow}
      ${transitions}
    </style>
    ${snowflakes}
  `
  document.body.appendChild(snowDiv)
}
