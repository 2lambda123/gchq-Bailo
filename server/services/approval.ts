import { Types } from 'mongoose'
import { getEntitiesForUser, getUserListFromEntityList, parseEntityList } from '../utils/entity'
import { ApprovalStates, Approval, Entity } from '../../types/interfaces'
import { DeploymentDoc } from '../models/Deployment'
import ApprovalModel, { ApprovalTypes, ApprovalCategory } from '../models/Approval'
import { VersionDoc } from '../models/Version'
import { reviewApproval } from '../templates/reviewApproval'
import { BadReq } from '../utils/result'
import { sendEmail } from '../utils/smtp'
import { getUserByInternalId } from './user'
import { UserDoc } from '../models/User'

export async function createDeploymentApprovals({
  version,
  deployment,
}: {
  version: VersionDoc
  deployment: DeploymentDoc
}) {
  const managers = await parseEntityList(version.metadata.contacts.manager)

  if (!managers.valid) {
    throw BadReq({ managers: version.metadata.contacts.manager }, `Invalid manager: ${managers.reason}`)
  }

  return createDeploymentApproval({
    approvers: version.metadata.contacts.manager,
    deployment,
    approvalType: ApprovalTypes.Manager,
  })
}

export async function createVersionApprovals({ version }: { version: VersionDoc }) {
  const [managers, reviewers] = await Promise.all([
    parseEntityList(version.metadata.contacts.manager),
    parseEntityList(version.metadata.contacts.reviewer),
  ])

  if (!managers.valid) {
    throw BadReq({ managers: version.metadata.contacts.manager }, `Invalid manager: ${managers.reason}`)
  }

  if (!reviewers.valid) {
    throw BadReq({ reviewers: version.metadata.contacts.reviewer }, `Invalid reviewer: '${reviewers.reason}'`)
  }

  const managerApproval = createVersionApproval({
    approvers: version.metadata.contacts.manager,
    version,
    approvalType: ApprovalTypes.Manager,
  })

  const reviewerApproval = createVersionApproval({
    approvers: version.metadata.contacts.reviewer,
    version,
    approvalType: ApprovalTypes.Reviewer,
  })

  return Promise.all([managerApproval, reviewerApproval])
}

async function createDeploymentApproval({
  approvers,
  approvalType,
  deployment,
}: {
  approvers: Array<Entity>
  approvalType: ApprovalTypes
  deployment: DeploymentDoc
}): Promise<Approval> {
  return createApproval({
    documentType: 'deployment',
    document: deployment,
    approvers,

    approvalType,
    approvalCategory: ApprovalCategory.Deployment,
  })
}

async function createVersionApproval({
  approvers,
  approvalType,
  version,
}: {
  approvers: Array<Entity>
  approvalType: ApprovalTypes
  version: VersionDoc
}): Promise<Approval> {
  return createApproval({
    documentType: 'version',
    document: version,
    approvers,

    approvalType,
    approvalCategory: ApprovalCategory.Upload,
  })
}

export type ApprovalDocumentType = 'version' | 'deployment'
async function createApproval({
  documentType,
  document,
  approvers,
  approvalType,
  approvalCategory,
}: {
  documentType: ApprovalDocumentType
  document: VersionDoc | DeploymentDoc
  approvers: Array<Entity>

  approvalType: ApprovalTypes
  approvalCategory: ApprovalCategory
}) {
  const doc = {
    [documentType]: document._id,
    approvers,

    approvalType,
    approvalCategory,
  }

  // If an approval already exists, we don't want to create a duplicate approval
  const { value: approval, lastErrorObject } = (await ApprovalModel.findOneAndUpdate(
    doc,
    {
      [documentType]: document._id,
      status: ApprovalStates.NoResponse,
    },
    {
      upsert: true,
      new: true,
      rawResult: true,
    }
  )) as unknown as any

  if (!lastErrorObject?.updatedExisting) {
    const users = await getUserListFromEntityList(approvers)

    for (const user of users) {
      if (user.email) {
        // we created a new approval, send out a notification.
        await sendEmail({
          to: user.email,
          ...reviewApproval({
            document,
            approvalCategory,
          }),
        })
      }
    }
  }

  return approval
}

export async function readNumApprovals({ userId }: { userId: Types.ObjectId }) {
  const user = await getUserByInternalId(userId)

  if (!user) {
    throw new Error(`Finding approvals for user that does not exist: ${userId}`)
  }

  const userEntities = await getEntitiesForUser(user)

  return ApprovalModel.countDocuments({
    status: 'No Response',
    $or: userEntities.map((userEntity) => ({
      approvers: { $elemMatch: { kind: userEntity.kind, id: userEntity.id } },
    })),
  })
}

export type ApprovalFilter = Types.ObjectId | undefined
export async function readApprovals({
  approvalCategory,
  filter,
  archived,
}: {
  approvalCategory: ApprovalCategory
  filter: ApprovalFilter
  archived: boolean
}) {
  const query: any = {
    status: 'No Response',
    approvalCategory,
  }

  if (filter) {
    const user = await getUserByInternalId(filter)

    if (!user) {
      throw new Error(`Finding approvals for user that does not exist: ${filter}`)
    }

    const userEntities = await getEntitiesForUser(user)

    query.$or = userEntities.map((userEntity) => ({
      approvers: { $elemMatch: { kind: userEntity.kind, id: userEntity.id } },
    }))
  }

  if (archived) {
    query.status = { $ne: 'No Response' }
  }

  return ApprovalModel.find(query)
    .populate({
      path: 'version',
      populate: { path: 'model' },
    })
    .populate({
      path: 'deployment',
      populate: { path: 'model' },
    })
    .sort({ updatedAt: -1 })
}

export async function getApproval({ approvalId }: { approvalId: string | Types.ObjectId }) {
  const approval = await ApprovalModel.findById(approvalId)

  if (!approval) {
    throw BadReq({ approvalId }, `Unable to find approval '${approvalId}'`)
  }

  return approval
}

export async function deleteApprovalsByVersion(user: UserDoc, version: VersionDoc) {
  return (ApprovalModel as any).delete(
    {
      version: version._id,
    },
    user._id
  )
}

export async function deleteApprovalsByDeployment(user: UserDoc, deployment: DeploymentDoc) {
  return (ApprovalModel as any).delete(
    {
      deployment: deployment._id,
    },
    user._id
  )
}
