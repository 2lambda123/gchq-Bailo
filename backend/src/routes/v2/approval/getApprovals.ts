import bodyParser from 'body-parser'
import { Request, Response } from 'express'
import { z } from 'zod'

import { ApprovalRequestInterface } from '../../../models/v2/Approval.js'
import { BadReq } from '../../../utils/v2/error.js'
import { parse } from '../../../utils/v2/validate.js'

export const getApprovalsSchema = z.object({
  query: z.object({
    isActive: z.string({
      required_error: 'Missing parameter isActive.',
    }),
  }),
})

interface GetApprovalsResponse {
  data: {
    approvals: Array<ApprovalRequestInterface>
  }
}

export const getApprovals = [
  bodyParser.json(),
  async (req: Request, res: Response<GetApprovalsResponse>) => {
    const _ = parse(req, getApprovalsSchema)
    const { isActive } = req.query
    if (isActive !== 'true' && isActive !== 'false') {
      throw BadReq('isActive should be true or false')
    }
    const active = req.query.isActive === 'true'
    return active
      ? res.json({
          data: {
            approvals: [
              {
                model: 'yolo',
                release: '3.0.2',
                kind: 'technical',
                isActive: true,
                createdAt: new Date('08/13/2023'),
                updatedAt: new Date('08/14/2023'),
              },
              {
                model: 'yolo',
                release: '3.0.1',
                kind: 'technical',
                isActive: true,
                createdAt: new Date('08/12/2023'),
                updatedAt: new Date('08/12/2023'),
              },
            ],
          },
        })
      : res.json({
          data: {
            approvals: [
              {
                model: 'yolo',
                release: '3.0.0',
                kind: 'technical',
                isActive: false,
                createdAt: new Date('08/11/2023'),
                updatedAt: new Date('08/11/2023'),
              },
            ],
          },
        })
  },
]
