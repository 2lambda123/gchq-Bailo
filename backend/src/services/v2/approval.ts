import { testReleaseInactiveApproval } from '../../../test/testUtils/testModels.js'
import authorisation from '../../connectors/v2/authorisation/index.js'
import Approval, { ApprovalInterface } from '../../models/v2/Approval.js'
import { UserDoc } from '../../models/v2/User.js'

export async function findApprovalsByActive(user: UserDoc, active: boolean): Promise<ApprovalInterface[]> {
  const approvals = await Approval.aggregate()
    .match({ active })
    .sort({ createdAt: -1 })
    // Populate model entries
    .lookup({ from: 'v2_models', localField: 'model', foreignField: 'id', as: 'model' })
    // Populate model as value instead of array
    .append({ $set: { model: { $arrayElemAt: ['$model', 0] } } })
    .match({
      'model.collaborators': {
        $elemMatch: {
          entity: { $in: await authorisation.getEntities(user) },
        },
      },
    })

  return approvals
}
export async function countApprovals(user: UserDoc): Promise<number> {
  return (await findApprovalsByActive(user, true)).length
}

/**
 * Added for testing- remove?
 */
export async function addDefaultApprovals() {
  const approvalDoc = new Approval(testReleaseInactiveApproval)

  await approvalDoc.save()
}
