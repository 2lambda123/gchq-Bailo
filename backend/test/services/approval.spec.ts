import { describe, expect, test, vi } from 'vitest'

import { countApprovals, findApprovalsByActive } from '../../src/services/v2/approval.js'

vi.mock('../../src/connectors/v2/authorisation/index.js', async () => ({
  default: { getEntities: vi.fn(() => ['user:test']) },
}))

const approvalModel = vi.hoisted(() => {
  const model: any = {}

  model.aggregate = vi.fn(() => model)
  model.match = vi.fn(() => model)
  model.sort = vi.fn(() => model)
  model.lookup = vi.fn(() => model)
  model.append = vi.fn(() => model)

  return model
})
vi.mock('../../src/models/v2/Approval.js', () => ({
  default: approvalModel,
}))

describe('services > approval', () => {
  test('findApprovalsByActive > active', async () => {
    const user: any = { dn: 'test' }
    await findApprovalsByActive(user, true)

    expect(approvalModel.match.mock.calls.at(0)).toMatchSnapshot()
    expect(approvalModel.match.mock.calls.at(1)).toMatchSnapshot()
  })

  test('findApprovalsByActive > not active', async () => {
    const user: any = { dn: 'test' }
    await findApprovalsByActive(user, false)

    expect(approvalModel.match.mock.calls.at(0)).toMatchSnapshot()
    expect(approvalModel.match.mock.calls.at(1)).toMatchSnapshot()
  })

  test('countApprovals > successful', async () => {
    const user: any = { dn: 'test' }
    await countApprovals(user)

    expect(approvalModel.match.mock.calls.at(0)).toMatchSnapshot()
    expect(approvalModel.match.mock.calls.at(1)).toMatchSnapshot()
  })
})
