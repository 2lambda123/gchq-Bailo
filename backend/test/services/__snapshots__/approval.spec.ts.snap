// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`services > approval > findApprovalsByActive > active 1`] = `
[
  {
    "active": true,
  },
]
`;

exports[`services > approval > findApprovalsByActive > active 2`] = `
[
  {
    "model.collaborators": {
      "$elemMatch": {
        "entity": {
          "$in": [
            "user:test",
          ],
        },
      },
    },
  },
]
`;

exports[`services > approval > findApprovalsByActive > not active 1`] = `
[
  {
    "active": false,
  },
]
`;

exports[`services > approval > findApprovalsByActive > not active 2`] = `
[
  {
    "model.collaborators": {
      "$elemMatch": {
        "entity": {
          "$in": [
            "user:test",
          ],
        },
      },
    },
  },
]
`;
